packages/data-service/src/graphqlObservable.ts

```diff
@@ -0,0 +1,309 @@
+}
+
+interface Schema extends GraphQLSchema {
+  _typeMap: TypeMap;
+}
+
+interface OperationNode {
+  operation: "query" | "mutation";
+}
+
+type SchemaNode = SelectionNode | DefinitionNode;
+
+function isOperationDefinition(node: any): node is OperationNode {
+  return node.kind === Kind.OPERATION_DEFINITION;
+}
+function isFieldNode(node: SchemaNode): node is FieldNode {
+  return node.kind === Kind.FIELD;
+}
+
+// We don't treat OperationDefinitions as Definitions but as entry points for our execution
+function isDefinitionNode(node: SchemaNode): node is DefinitionNode {
+  return (
+    node.kind === Kind.FRAGMENT_DEFINITION ||
+    isTypeSystemDefinitionNode(node) ||
+    isTypeSystemExtensionNode(node)
+  );
+}
+
+interface FieldWithResolver extends GraphQLField<any, any> {
+  resolve: GraphQLFieldResolver<any, any, any>;
+}
+
+function isFieldWithResolver(
+  field: GraphQLField<any, any>
+): field is FieldWithResolver {
+  return field.resolve instanceof Function;
+}
+
+export function graphqlObservable<T = object>(
+  doc: DocumentNode,
+  schema: Schema,
+  context: object
+): Observable<{ data: T }> {
+  if (doc.definitions.length !== 1) {
+    return throwObservable("document root must have a single definition");
+  }
+  const types = schema._typeMap;
+
+  return resolve(doc.definitions[0], context, null, null).map((data: T) => ({
+    data
+  }));
+
+  function resolve(
+    definition: SchemaNode,
+    context: object,
+    parent: any,
+    type: GraphQLType | null
+  ) {
+    if (isOperationDefinition(definition)) {
+      const nextType = getResultType(type, definition, parent);
+
+      return resolveResult(definition, context, null, nextType);
+    }
+
+    // The definition gives us the field to resolve
+    if (isFieldNode(definition)) {
+      const field = getField(type, definition);
+
+      // Something unexpcected was passed into getField
+      if (field === null) {
+        // TODO: find better worded error message
+        return throwObservable(
+          `field was not of the right type. Given type: ${type}`
+        );
+      }
+
+      const resolvedObservable = resolveField(
+        field,
+        definition,
+        context,
+        parent
+      );
+
+      // Directly return the leaf nodes
+      if (definition.selectionSet === undefined) {
+        return resolvedObservable;
+      }
+
+      return resolvedObservable.concatMap(emitted => {
+        if (!emitted) {
+          return throwObservable("resolver emitted empty value");
+        }
+
+        if (emitted instanceof Array) {
+          return resolveArrayResults(definition, context, emitted, type);
+        }
+
+        const nextType = getResultType(type, definition, emitted);
+        return resolveResult(definition, context, emitted, nextType);
+      });
+    }
+
+    // It is no operationDefinitionand no fieldNode, so it seems like an error
+    // TODO: throw better error here
+    return throwObservable("Input was strange");
```

Comment:
Can we just say `Input doesn't look like operationDefinitionand nor fieldNode`?

Reply:
yes, I completely overlooked this message when going through the code again, thanks for pointing it out 👍

Reply:
done 👍